frame/session/src/historical/shared.rs

```diff
@@ -0,0 +1,16 @@
+use super::*;
+
+pub(super) const PREFIX: &[u8] = b"historical";
+pub(super) const LAST_PRUNE: &[u8] = b"historical_last_prune";
+
+/// Derive the key used to store the list of validators
+pub(super) fn derive_key<P: AsRef<[u8]>>(prefix: P, session_index: SessionIndex) -> Vec<u8> {
+    let prefix: &[u8] = prefix.as_ref();
+    let encoded_session_index = session_index.encode();
```

Comment:
nit: you can use `using_encoded(|s| ...)` to avoid an allocation here, computing `concatenated` within the closure

Reply:
I could, but then I could not pre-allocate correctly, not sure which is worse

Reply:
```rust    
    let prefix: &[u8] = prefix.as_ref();
    session_index.using_encoded(|encoded_session_index| {
        assert!(encoded_session_index.len() > 0);
        let mut concatenated = Vec::with_capacity(prefix.len() + 1 + encoded_session_index.len());
        concatenated.extend_from_slice(prefix);
        concatenated.push('/' as u8);
        concatenated.extend_from_slice(encoded_session_index.as_slice());
        concatenated
    })
```

I would have thought this works, and it's what I meant by "computing `concatenated` within the closure"

Reply:
Did not think of moving all into the closure, will give it a shot

Reply:
Also a nit: `core::iter::chain` preserves the `ExactSizeIterator` property, which should mean that preallocation is unnecessary when `collect`ing from a simple `slice_a.iter().chain(b'/').chain(slice_b.iter())` as it will preallocate in `Vec::collect` using `size_hint`.